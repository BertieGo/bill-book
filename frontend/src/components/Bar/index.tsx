import React from "react";
import {Col, Row, Tooltip} from "antd";
import cloneDeep from "lodash/cloneDeep";
import isEqual from 'lodash/isEqual';

import {IChartProps, IPathItem} from "../../declare";
import {fixedFloat, floorFloat, getPercent} from "../../utils/math";

export default class Bar extends React.Component<IChartProps, {}>{

    shouldComponentUpdate(nextProps: Readonly<IChartProps>, nextState: Readonly<{}>, nextContext: any): boolean {
        return !isEqual(nextProps.data, this.props.data);
    }

    makeRowData = () => {
        const { data } = this.props;
        let dataCopy = cloneDeep(data);
        if (data.length > 0) {
            dataCopy = dataCopy.sort((prev, next) => next.amount - prev.amount);
            const maxPercent = dataCopy[0].percent;
            dataCopy.forEach(item => {
                item.barPercent = getPercent(item.percent, maxPercent)
            })
        }
        return dataCopy;
    }

    renderRow = (item: IPathItem) => {

        const displayPercent = `${floorFloat(item.percent)}%`;
        const displayAmount = fixedFloat(item.amount);

        return (
            <Row>
                <Tooltip title={`金额: ${displayAmount} 占比：${displayPercent}`}>
                    <Col span={20}>
                        <div className="row-bar" style={{ width: `${item.barPercent}%`, backgroundColor: item.color }}/>
                    </Col>
                </Tooltip>
                <Col span={4}>
                    <span className="ml20">{ displayAmount }</span>
                </Col>
            </Row>
        )
    }

    renderRowChart = (item: IPathItem) => {
        return (
            <Row className="mb10" key={item._id}>
                <Col span={4}>{ item.title }：</Col>
                <Col span={8}>{ this.renderRow(item) }</Col>
            </Row>
        )
    }

    render() {
        const data = this.makeRowData();
        return (
            <div>
                {
                    data.map(this.renderRowChart)
                }
            </div>
        )
    }
}
