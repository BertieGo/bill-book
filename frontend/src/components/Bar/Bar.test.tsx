import React from 'react';
import Bar from '../Bar';
import { render } from '@testing-library/react';

// test('Bar component', () => {
//     const data = [{
//         amount: 2672740,
//         category: '8s0p77c323',
//         color:  "#72a935",
//         percent: 55.4886,
//         time: "2019-06-30T16:00:00.000Z",
//         title: '支出',
//         type: 0,
//     }];
//
//     const component = render(
//         <Bar data={data as any[]}/>,
//     );
//     let tree = component;
//
//     console.log(tree)
//     // expect(tree).toMatchSnapshot()
// })


