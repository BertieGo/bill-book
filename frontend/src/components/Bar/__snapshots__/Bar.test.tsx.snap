// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT-TEST: Bar 1`] = `
<div>
  <div
    className="ant-row mb10"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-4"
      style={Object {}}
    >
      支出
      ：
    </div>
    <div
      className="ant-col ant-col-8"
      style={Object {}}
    >
      <div
        className="ant-row"
        style={Object {}}
      >
        <div
          className="ant-col ant-col-20"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <div
            className="row-bar"
            style={
              Object {
                "backgroundColor": "#72a935",
                "width": "100%",
              }
            }
          />
        </div>
        <div
          className="ant-col ant-col-4"
          style={Object {}}
        >
          <span
            className="ml20"
          >
            2672740
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`COMPONENT-TEST: Pie 1`] = `
<div>
  <div
    className="ant-row mb10"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-4"
      style={Object {}}
    >
      支出
      ：
    </div>
    <div
      className="ant-col ant-col-8"
      style={Object {}}
    >
      <div
        className="ant-row"
        style={Object {}}
      >
        <div
          className="ant-col ant-col-20"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <div
            className="row-bar"
            style={
              Object {
                "backgroundColor": "#72a935",
                "width": "100%",
              }
            }
          />
        </div>
        <div
          className="ant-col ant-col-4"
          style={Object {}}
        >
          <span
            className="ml20"
          >
            2672740
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Link renders correctly 1`] = `
<div>
  <div
    className="ant-row mb10"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-4"
      style={Object {}}
    >
      支出
      ：
    </div>
    <div
      className="ant-col ant-col-8"
      style={Object {}}
    >
      <div
        className="ant-row"
        style={Object {}}
      >
        <div
          className="ant-col ant-col-20"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <div
            className="row-bar"
            style={
              Object {
                "backgroundColor": "#72a935",
                "width": "100%",
              }
            }
          />
        </div>
        <div
          className="ant-col ant-col-4"
          style={Object {}}
        >
          <span
            className="ml20"
          >
            2672740
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
